### PR TITLE
Reliability and packet scheduling

### PR DESCRIPTION
**Pull request by sakshamagarwals**
Added retransmission mechanism
Added receiver-based scheduling
Added sender-based scheduling
Added RTT measurement

### CODE REVIEW

File: homatransport/src/application/WorkloadSynthesizer.h [hunk 1]
```diff
@@ -74,6 +74,8 @@ class WorkloadSynthesizer : public cSimpleModule
     int nextDestHostId; // -1 means the destination must be chosen
                         // randomely based on config.xml information.
     std::unordered_map<int, inet::L3Address> hostIdAddrMap;
+    //also need host address to host id map for logging purposes
```
Comment (hwassel) [inline homatransport/src/application/WorkloadSynthesizer.h]: Space afer //

File: homatransport/src/common/Lagger.cc [hunk 2]
```diff
@@ -49,8 +54,18 @@ Lagger::handleMessage(cMessage *msg)
     } else {
         if (hookType == "InputHook") {
             inputHookPktHandler(msg);
+            if(logPacketEvents){
```
Comment (hwassel) [inline homatransport/src/common/Lagger.cc]: space after if

File: homatransport/src/common/Lagger.cc [hunk 3]
```diff
@@ -49,8 +54,18 @@ Lagger::handleMessage(cMessage *msg)
...
+                cModule* parentHost = this->getParentModule();
+                cModule* grandParentHost = parentHost->getParentModule();
+                logFile << simTime() << " At input lagger: " << parentHost->getName() << " " << grandParentHost->getName() << " " << grandParentHost->getIndex() << std::endl;
+            }
         } else if (hookType == "OutputHook") {
             outputHookPktHandler(msg);
+            if(logPacketEvents){
```
Comment (hwassel) [inline homatransport/src/common/Lagger.cc]: Same.

File: homatransport/src/transport/VectioTransport.h [hunk 4]
```diff
@@ -53,25 +62,48 @@ class VectioTransport : public cSimpleModule
...
      */
     enum SelfMsgKind
     {
         START = 1,  // Timer type when the transport is in initialization phase.
-        STOP  = 2   // Timer type when the transport is in cleaning phase.
+        STOP  = 2,   // Timer type when the transport is in cleaning phase.
+        IBGRANTQUEUE = 3, // Timer type when the transport wants to process the 
+                       // grant queue
+        OBGRANTQUEUE = 4, // Timer type when the transport wants to process the 
+                       // grant queue
```
Comment (hwassel) [inline homatransport/src/transport/VectioTransport.h:88]: These two values are hard to read.

File: inet/src/inet/common/queue/PassiveQueueBase.cc [hunk 5]
```diff
@@ -90,8 +90,8 @@ void PassiveQueueBase::handleMessage(cMessage *msg)
                 case PktType::UNSCHED_DATA:
                     emit(unschedDataQueueingTimeSignal, SIMTIME_ZERO);
                     break;
-                default:
-                    throw cRuntimeError("HomaPkt arrived at the queue has unknown type.");
+                // default:
```
Comment (hwassel) [inline inet/src/inet/common/queue/PassiveQueueBase.cc]: remove commented lines. What happens now that there is no default?

File: inet/src/inet/common/queue/PassiveQueueBase.cc [hunk 6]
```diff
@@ -149,8 +149,8 @@ void PassiveQueueBase::requestPacket()
                 case PktType::UNSCHED_DATA:
                     emit(unschedDataQueueingTimeSignal, simTime() - msg->getArrivalTime());
                     break;
-                default:
-                    throw cRuntimeError("HomaPkt arrived at the queue has unknown type.");
+                // default:
+                //     throw cRuntimeError("HomaPkt arrived at the queue has unknown type.");
```
Comment (hwassel) [inline inet/src/inet/common/queue/PassiveQueueBase.cc]: Same here.

File: inet/src/inet/linklayer/ethernet/EtherMACFullDuplex.cc [hunk 7]
```diff
@@ -444,8 +444,8 @@ EtherMACFullDuplex::countHomaPktBytes(EtherFrame* curFrame, bool isSent)
                     homaBytesCounter.numUnschedPktRecvOk++;
                 }
                 break;
-            default:
-                throw cRuntimeError("HomaPkt arrived at the queue has unknown type.");
+            // default:
```
Comment (hwassel) [inline inet/src/inet/linklayer/ethernet/EtherMACFullDuplex.cc]: And here.

File: inet/src/inet/common/queue/DropTailQueue.cc [hunk 8]
```diff
@@ -97,6 +101,36 @@ cMessage *DropTailQueue::enqueue(cMessage *msg)
     cPacket* pkt = check_and_cast<cPacket*>(msg);
     queue.insert(pkt);
 
+    if(logPacketEvents){
```
Comment (hwassel) [inline inet/src/inet/common/queue/DropTailQueue.cc:104]: The following code is repeated few times. Can you make it a utility function?

**Review by hwassel — COMMENTED** [review]
I have a high level comment. Are you planning to unit test vectio transport? 
I am still going through the bulk of the CL.

File: inet/src/inet/common/queue/DropTailQueue.cc (hunk 8 above)
Comment (sakshamagarwals) [reply inet/src/inet/common/queue/DropTailQueue.cc:104]: The logging inside these if conditions are different depending upon the place of logging. I'll however try to see if if I can make this a utility function across the code.

**Pull request merged**